### ISSUE TITLE
[FEAT][ELEMENTWISE] infrastructure: template base classes, smoke test, and strategy benchmark

### ISSUE THREAD

**Issue by lcy-seso** [tile-ai/TileOPs#436]
## 1. Feature Scope

Build the complete template infrastructure for 66 elementwise ops:
- 3 kernel template base classes: `UnaryKernel`, `BinaryKernel`, `FusedGatedKernel`
- 3 kernel strategies per unary (direct, explicit_parallel, register_copy), 2 per binary (direct, explicit_parallel)
- `coalesce_broadcast_dims` utility for N-dim broadcast
- 3 Op template base classes: `UnaryOp`, `BinaryOp`, `FusedGatedOp`
- 1 example concrete op per template: relu, add, silu_and_mul
- Strategy benchmark → determine `DEFAULT_STRATEGY`

**Not in scope**: remaining 63 ops (#437 #438 #439), torch.compile (#440), fp8 (#442).

### Files

| File | Action |
|------|--------|
| `tileops/kernels/elementwise.py` | New — template base classes + strategy factories + 3 example ops |
| `tileops/ops/elementwise.py` | New — Op base classes + coalesce utility + 3 example ops |
| `tileops/kernels/__init__.py` | Add re-exports: `UnaryKernel`, `BinaryKernel`, `FusedGatedKernel` only |
| `tileops/ops/__init__.py` | Add re-exports: `UnaryOp`, `BinaryOp`, `FusedGatedOp` only |
| `tests/ops/test_activation.py` | New — relu smoke test |
| `tests/ops/test_binary_arith.py` | New — add smoke test + broadcast tests |
| `tests/ops/test_fused_gated.py` | New — silu_and_mul smoke test |
| `benchmarks/ops/bench_activation.py` | New — unary strategy benchmark |
| `benchmarks/ops/bench_binary_arith.py` | New — binary strategy benchmark |

### 2.1 Kernel Strategies

All elementwise ops: **no shared memory**, pure `Global → Register → Compute → Register → Global`.

`num_per_thread` is dtype-aware: fp32→4, fp16/bf16→8 (aligned to 128-bit `uint4` load).

#### Strategy 1: `direct`

```python
def _make_unary_direct(N, dtype, op_func, threads=256):
    @T.prim_func
    def main(x: T.Tensor((N,), dtype), y: T.Tensor((N,), dtype)):
        with T.Kernel(T.ceildiv(N, threads), threads=threads) as bx:
            for i in T.Parallel(threads):
                y[bx * threads + i] = op_func(x[bx * threads + i])
    return main
```

#### Strategy 2: `explicit_parallel` (default)

```python
def _make_unary_explicit(N, dtype, op_func, threads=256, num_per_thread=8):
    block_size = threads * num_per_thread
    @T.prim_func
    def main(x: T.Tensor((N,), dtype), y: T.Tensor((N,), dtype)):
        with T.Kernel(T.ceildiv(N, block_size), threads=threads) as bx:
            for i, j in T.Parallel(threads, num_per_thread):
                idx = (bx * threads + i) * num_per_thread + j
                y[idx] = op_func(x[idx])
    return main
```

#### Strategy 3: `register_copy`

```python
def _make_unary_regcopy(N, dtype, op_func, threads=256, num_per_thread=8):
    block_size = threads * num_per_thread
    @T.prim_func
    def main(x: T.Tensor((N,), dtype), y: T.Tensor((N,), dtype)):
        with T.Kernel(T.ceildiv(N, block_size), threads=threads) as bx:
            x_reg = T.alloc_fragment((block_size,), dtype)
            y_reg = T.alloc_fragment((block_size,), dtype)
            T.copy(x[bx * block_size:(bx + 1) * block_size], x_reg)
            for i, j in T.Parallel(threads, num_per_thread):
                y_reg[i * num_per_thread + j] = op_func(x_reg[i * num_per_thread + j])
            T.copy(y_reg, y[bx * block_size:(bx + 1) * block_size])
    return main
```

#### Binary: `explicit_parallel` with stride-based N-dim broadcast

```python
def _make_binary_explicit(N_total, dtype, op_func,
                          coalesced_shape, a_strides, b_strides,
                          a_numel, b_numel,
                          output_dtype=None, threads=256, num_per_thread=8):
    block_size = threads * num_per_thread
    ndim = len(coalesced_shape)
    out_dtype = output_dtype or dtype
    # Precompute divisors: divisors[d] = product(coalesced_shape[d+1:])
    divisors = [1] * ndim
    for i in range(ndim - 2, -1, -1):
        divisors[i] = divisors[i + 1] * coalesced_shape[i + 1]

    @T.prim_func
    def main(a: T.Tensor((a_numel,), dtype),
             b: T.Tensor((b_numel,), dtype),
             y: T.Tensor((N_total,), out_dtype)):
        with T.Kernel(T.ceildiv(N_total, block_size), threads=threads) as bx:
            for i, j in T.Parallel(threads, num_per_thread):
                flat_idx = (bx * threads + i) * num_per_thread + j
                a_off, b_off, remaining = 0, 0, flat_idx
                for d in range(ndim - 1):
                    coord = remaining // divisors[d]
                    remaining = remaining % divisors[d]
                    a_off += coord * a_strides[d]
                    b_off += coord * b_strides[d]
                a_off += remaining * a_strides[-1]
                b_off += remaining * b_strides[-1]
                y[flat_idx] = op_func(a[a_off], b[b_off])
    return main
```

Binary `direct` strategy: same logic, `T.Parallel(threads)` with 1 element per thread.

#### FusedGated kernel

```python
def _make_fused_gated_kernel(M, N, dtype, activation_func, threads=256, npt=8):
    block_N = threads * npt
    @T.prim_func
    def main(x: T.Tensor((M, 2 * N), dtype), y: T.Tensor((M, N), dtype)):
        with T.Kernel(T.ceildiv(N, block_N), M, threads=threads) as (bx, by):
            for i, j in T.Parallel(threads, npt):
                col = (bx * threads + i) * npt + j
                gate = x[by, col]
                value = x[by, N + col]
                y[by, col] = activation_func(gate) * value
    return main
```

#### UnaryKernel

```python
class UnaryKernel(Kernel):
    supported_archs = [80, 86, 89, 90]
    STRATEGIES = ["direct", "explicit_parallel", "register_copy"]
    DEFAULT_STRATEGY = "explicit_parallel"  # Updated by benchmark

    @staticmethod
    def op_func(x):
        raise NotImplementedError

    def __init__(self, N_total, dtype, strategy=None, config=None, tune=False):
        super().__init__()
        self.N_total = N_total
        self.dtype = dtype
        strategy = strategy or self.DEFAULT_STRATEGY
        self.kernel = self._build_kernel(strategy)  # ① build kernel first
        self.init_config(config, tune)               # ② then config/autotune

    def _build_kernel(self, strategy):
        cfg = self.default_config
        if strategy == "direct":
            return _make_unary_direct(self.N_total, self.dtype_str, self.op_func, cfg["threads"])
        elif strategy == "explicit_parallel":
            return _make_unary_explicit(self.N_total, self.dtype_str, self.op_func, cfg["threads"], cfg["num_per_thread"])
        elif strategy == "register_copy":
            return _make_unary_regcopy(self.N_total, self.dtype_str, self.op_func, cfg["threads"], cfg["num_per_thread"])

    @property
    def default_config(self):
        npt = 4 if self.dtype == torch.float32 else 8
        return {"threads": 256, "num_per_thread": npt}

    def forward(self, x):
        return self.kernel(x)
```

#### BinaryKernel

```python
class BinaryKernel(Kernel):
    supported_archs = [80, 86, 89, 90]
    STRATEGIES = ["direct", "explicit_parallel"]  # No register_copy (incompatible with stride-based access)
    DEFAULT_STRATEGY = "explicit_parallel"
    OUTPUT_DTYPE = None  # Subclass override: e.g. EqKernel.OUTPUT_DTYPE = "bool"

    @staticmethod
    def op_func(a, b):
        raise NotImplementedError

    def __init__(self, N_total, dtype, coalesced_shape, a_strides, b_strides,
                 a_numel, b_numel, strategy=None, config=None, tune=False):
        super().__init__()
        self.N_total, self.dtype = N_total, dtype
        self.coalesced_shape, self.a_strides, self.b_strides = coalesced_shape, a_strides, b_strides
        self.a_numel, self.b_numel = a_numel, b_numel
        strategy = strategy or self.DEFAULT_STRATEGY
        self.kernel = self._build_kernel(strategy)
        self.init_config(config, tune)

    def _build_kernel(self, strategy):
        cfg = self.default_config
        if strategy == "direct":
            return _make_binary_direct(self.N_total, self.dtype_str, self.op_func,
                self.coalesced_shape, self.a_strides, self.b_strides,
                self.a_numel, self.b_numel, output_dtype=self.OUTPUT_DTYPE, threads=cfg["threads"])
        elif strategy == "explicit_parallel":
            return _make_binary_explicit(self.N_total, self.dtype_str, self.op_func,
                self.coalesced_shape, self.a_strides, self.b_strides,
                self.a_numel, self.b_numel, output_dtype=self.OUTPUT_DTYPE,
                threads=cfg["threads"], num_per_thread=cfg["num_per_thread"])

    @property
    def default_config(self):
        npt = 4 if self.dtype == torch.float32 else 8
        return {"threads": 256, "num_per_thread": npt}

    def forward(self, a, b):
        return self.kernel(a, b)
```

#### FusedGatedKernel

```python
class FusedGatedKernel(Kernel):
    supported_archs = [80, 86, 89, 90]

    @staticmethod
    def activation_func(x):
        raise NotImplementedError

    def __init__(self, M, N, dtype, config=None, tune=False):
        super().__init__()
        self.M, self.N, self.dtype = M, N, dtype
        cfg = self.default_config
        self.kernel = _make_fused_gated_kernel(M, N, self.dtype_str, self.activation_func, cfg["threads"], cfg["num_per_thread"])
        self.init_config(config, tune)

    @property
    def default_config(self):
        npt = 4 if self.dtype == torch.float32 else 8
        return {"threads": 256, "num_per_thread": npt}

    def forward(self, x):
        return self.kernel(x)
```

### 2.3 Example Concrete Ops

```python
# Kernel layer
class ReluKernel(UnaryKernel):
    @staticmethod
    def op_func(x):  return T.max(x, 0)

class AddKernel(BinaryKernel):
    @staticmethod
    def op_func(a, b):  return a + b

class SiluAndMulKernel(FusedGatedKernel):
    @staticmethod
    def activation_func(x):  return x * T.sigmoid(x)
```

```python
# Op layer
class ReluOp(UnaryOp):
    _op_name = "relu"
    kernel_cls = ReluKernel

class AddOp(BinaryOp):
    _op_name = "add"
    kernel_cls = AddKernel
```

### 2.4 `coalesce_broadcast_dims` (Op layer utility)

```python
def coalesce_broadcast_dims(a_shape, b_shape):
    """Coalesce N-dim broadcast into minimal effective dimensions.

    Returns: (out_shape, coalesced_shape, a_strides, b_strides)
    """
    out_shape = torch.broadcast_shapes(a_shape, b_shape)
    ndim = len(out_shape)
    a_pad = (1,) * (ndim - len(a_shape)) + tuple(a_shape)
    b_pad = (1,) * (ndim - len(b_shape)) + tuple(b_shape)

    def _make_strides(padded_shape):
        strides = [1] * ndim
        for i in range(ndim - 2, -1, -1):
            strides[i] = strides[i + 1] * padded_shape[i + 1]
        return [0 if padded_shape[i] == 1 else strides[i] for i in range(ndim)]

    a_raw, b_raw = _make_strides(a_pad), _make_strides(b_pad)

    groups = [(out_shape[0], a_raw[0], b_raw[0])]
    for i in range(1, ndim):
        prev_out, prev_as, prev_bs = groups[-1]
        a_can = (a_raw[i] == 0 and prev_as == 0) or (a_raw[i] != 0 and prev_as == a_raw[i] * out_shape[i])
        b_can = (b_raw[i] == 0 and prev_bs == 0) or (b_raw[i] != 0 and prev_bs == b_raw[i] * out_shape[i])
        if a_can and b_can:
            groups[-1] = (prev_out * out_shape[i], a_raw[i], b_raw[i])
        else:
            groups.append((out_shape[i], a_raw[i], b_raw[i]))

    groups = [g for g in groups if g[0] > 1] or [(1, 0, 0)]
    coalesced_shape = tuple(g[0] for g in groups)
    a_strides = tuple(g[1] for g in groups)
    b_strides = tuple(g[2] for g in groups)
    return out_shape, coalesced_shape, a_strides, b_strides
```

### 2.5 Op Template Base Classes

```python
class UnaryOp(Op):
    kernel_cls: type[UnaryKernel]

    def __init__(self, N_total, dtype, kernel_map=None, tune=False):
        self.N_total, self.dtype = N_total, dtype
        self.dispatch_kernel(kernel_map)
        self.kernel = self.kernel_map[self._op_name](N_total, dtype, tune=tune)

    @property
    def default_kernel_map(self):
        return {self._op_name: self.kernel_cls}

    def forward(self, x):
        orig_shape = x.shape
        x = x.contiguous().reshape(-1)
        return self.kernel(x).reshape(orig_shape)


class BinaryOp(Op):
    kernel_cls: type[BinaryKernel]

    def __init__(self, a_shape, b_shape, dtype, kernel_map=None, tune=False):
        self.dtype = dtype
        out_shape, coalesced_shape, a_strides, b_strides = coalesce_broadcast_dims(a_shape, b_shape)
        self.out_shape = out_shape
        self.N_total = int(torch.tensor(out_shape).prod().item())
        self.a_numel = int(torch.tensor(a_shape).prod().item())
        self.b_numel = int(torch.tensor(b_shape).prod().item())
        self.dispatch_kernel(kernel_map)
        self.kernel = self.kernel_map[self._op_name](
            self.N_total, dtype, coalesced_shape, a_strides, b_strides,
            self.a_numel, self.b_numel, tune=tune)

    @property
    def default_kernel_map(self):
        return {self._op_name: self.kernel_cls}

    def forward(self, a, b):
        return self.kernel(a.contiguous().view(-1), b.contiguous().view(-1)).reshape(self.out_shape)
```

### 2.6 Strategy & Config Design Notes

- `strategy` is a constructor param, NOT part of `config`. Different strategies produce different kernel functions; `config` tunes params within the same kernel.
- `DEFAULT_STRATEGY` starts as `"explicit_parallel"`. **Updated after benchmark** (this issue).
- `config` = `{threads, num_per_thread}`. No `autotune_configs` in first phase.
- Boundary checks handled automatically by TileLang `LegalizeSafeMemoryAccess`.

### Smoke Tests (L1, 3 cases)

| Op | Dtype | Size | Broadcast |
|----|-------|------|-----------|
| relu | fp16 | 1M | same-shape |
| add | fp16 | 1M | same-shape |
| silu_and_mul | fp16 | 1M | N/A (2D input) |

Reference: PyTorch equivalent in fp32, cast back.

### Broadcast Tests (L3, 4 cases, add only)

| Pattern | Shapes | Coalesced ndim | Divmod count |
|---------|--------|----------------|-------------|
| bias-add | `(B,S,D) + (1,1,D)` | 2 | 1 |
| row broadcast | `(B,S,D) + (B,S,1)` | 2 | 1 |
| scalar | `(M,N) + (1,1)` | 2 | 1 |
| interleaved | `(A,1,C) + (1,B,1)` | 3 | 2 |

### coalesce_broadcast_dims Unit Tests

Verify coalesce output for all 6 patterns in the table above + same-shape (ndim=1) + outer product.

### Strategy Benchmark

See #445 for the full benchmark plan. This issue implements the benchmark harness and runs it.

Tolerance: fp32 atol=1e-5 rtol=1e-5, fp16 atol=1e-3 rtol=1e-3, bf16 atol=1.6e-2 rtol=1.6e-2.

## 4. Risks

| Risk | Mitigation |
|------|------------|
| `explicit_parallel` may not be optimal default | Benchmark all strategies, update `DEFAULT_STRATEGY` |
| stride-based binary offset may prevent vectorized loads | R1 benchmark (#445) — add fast paths if needed |
| TileLang boundary guard may hurt tail block perf | R5 benchmark (#445) — consider padding if >10% drop |
| JIT compilation overhead for varying shapes | R3 benchmark (#445) |

## 5. Acceptance Criteria

- [ ] UnaryKernel, BinaryKernel, FusedGatedKernel implemented with all strategies
- [ ] `coalesce_broadcast_dims` handles all 6+ broadcast patterns correctly
- [ ] relu, add, silu_and_mul pass L1 correctness tests
- [ ] 4 broadcast patterns pass L3 tests
- [ ] Strategy benchmark run; `DEFAULT_STRATEGY` updated with rationale
- [ ] `__init__.py` re-exports base classes only (not concrete ops)
